usr/src/uts/common/fs/zfs/zap_micro.c

```diff
@@ -527,10 +526,12 @@ zap_lockdir(objset_t *os, uint64_t obj, dmu_tx_t *tx,
 			dprintf("upgrading obj %llu: num_entries=%u\n",
 			    obj, zap->zap_m.zap_num_entries);
 			*zapp = zap;
-			return (mzap_upgrade(zapp, tx, 0));
+			int err = mzap_upgrade(zapp, tag, tx, 0);
+			if (err != 0)
+				zap_unlockdir(zap, tag);
```

Comment:
`zap_lockdir` only rele's the dbuf if `zap_lockdir_impl()` returns 0, so it won't in this case.